### PR TITLE
Jon/bit pay v2

### PR DESCRIPTION
**Pull request by Jon-edge**
#### PR Requirements

If you have made **any** visual changes to the GUI. Make sure you have:
- [ x] Tested on iOS Tablet
- [ x] Tested on small Android
- [ x] n/a
- 
Simulator Screen Shot - iPhone 11 Pro Max - 2022-01-25 at 10 29 42 Simulator Screen Shot - iPhone 11 Pro Max - 2022-01-25 at 10 31 14

Uploading Screen Recording 2022-01-25 at 10.27.15 AM.mov…

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201621837365742

### CODE REVIEW

File: src/actions/BitPayActions.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  // Pass spend info to the send scene for user to confirm tx
+  const spendTarget = spendInfo.spendTargets[0]
+  const guiMakeSpendInfo = {
+    currencyCode,
+    nativeAmount: spendTarget.nativeAmount,
+    publicAddress: spendTarget.publicAddress,
+    networkFeeOption: 'standard',
+    metadata: {
+      name: sprintf(s.strings.bitpay_metadata_name, paymentId),
+      notes: `${paymentId}`
```
Comment (peachbits) [inline src/actions/BitPayActions.js:164]: unnecessary string literal

File: src/util/DeepLinkParser.js [hunk 2]
```diff
@@ -26,7 +26,19 @@ export function parseDeepLink(uri: string, opts: { aztecoApiKey?: string } = {})
     return parseDownloadLink(url)
   }
 
-  // Handle the edge:// protocol:
+  // Handle bitpay.com links.
+  // We always want to bypass the plugin, even if a scheme (i.e. bitcoin:) is
+  // defined because it is valid for the user to accept any supported currency
+  // besides the specific currency defined in the uri's scheme.
+  if (uri.includes('https:') && uri.includes('bitpay.com')) {
+    if (url.protocol !== 'https:') {
```
Comment (peachbits) [inline src/util/DeepLinkParser.js:34]: Is it possible that the uri includes 'https:' but the url protocol wouldn't also be 'https:'? If it is, add examples of acceptable bitpay urls as comments

File: android/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -62,6 +62,8 @@
         <data android:scheme="https" />
         <data android:host="deep.edge.app" />
         <data android:host="dl.edge.app" />
+        <data android:host="test.bitpay.com" />
```
Comment (peachbits) [inline android/app/src/main/AndroidManifest.xml:65]: A similar change needs to be made in ios/edge/edge.entitlements

**Review by peachbits — CHANGES_REQUESTED** [review]

File: src/types/BitPayError.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  MultiInstructionInvoice: 'MultiInstructionInvoice',
+  TxVerificationMismatch: 'TxVerificationMismatch'
+}
+
+/**
+ * @internal
+ * Internal Mapping object from BitPayErrorCode to a BitPayErrorHandler
+ */
+const HandlersByCode = {
+  [BitPayErrorCode.EmptyOutputInvoice]: (params: { uri: string }) => s.strings.error_bitpay_empty_output_invoice,
```
[the remark this answers is not in the document]
Comment (Jon-edge) [reply src/types/BitPayError.js:34]: The idea here was for data to be stored in this error object for debugging, separate and more verbose from what is actually shown to the user on the airship. I can add an env setting for this, but I wanted to start some incremental changes towards error handling in this feature.

File: src/actions/BitPayActions.js [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    body: JSON.stringify({
+      chain: requestCurrencyCode,
+      currency: requestCurrencyCode
+    })
+  })
+
+  // Validate invoice data
+  const invoiceResponse = asBpInvoiceResponse(responseJson)
+  if (invoiceResponse.instructions.length > 1) {
+    throw new BitPayError(BitPayErrorCode.MultiInstructionInvoice, { uri })
```
[the remark this answers is not in the document]
Comment (Jon-edge) [reply src/actions/BitPayActions.js:106]: See comment in https://github.com/EdgeApp/edge-react-gui/pull/3115#discussion_r792065667

File: src/constants/WalletAndCurrencyConstants.js [hunk 6]
```diff
@@ -351,7 +363,8 @@ export const SPECIAL_CURRENCY_INFO: {
       privateKeyInstructions: s.strings.create_wallet_import_input_key_or_seed_instructions
     },
     isCustomTokensSupported: true,
-    isRbfSupported: true
+    isRbfSupported: true,
+    isBitPayProtocolSupported: true
```
[the remark this answers is not in the document]
Comment (Jon-edge) [reply src/constants/WalletAndCurrencyConstants.js:367]: Not in this PR

File: src/util/DeepLinkParser.js (hunk 2 above)
Comment (Jon-edge) [reply src/util/DeepLinkParser.js:34]: This case is catching a uri scheme representing a coin like "bitcoin:" that would be stored in the url.protocol param.
I can clarify the comment.

File: android/app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (Jon-edge) [reply android/app/src/main/AndroidManifest.xml:65]: Seems to work fine without any changes when simulating deep links via terminal or Safari

File: src/types/BitPayError.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (peachbits) [reply src/types/BitPayError.js:34]: You can remove the unused params from the function declarations in HandlersByCode it won't affect the ability to save the data to the error object. These are functionally identical as they both return the same string regardless of the arguments passed:

[BitPayErrorCode.EmptyOutputInvoice]: (params: { uri: string }) => s.strings.error_bitpay_empty_output_invoice
[BitPayErrorCode.EmptyOutputInvoice]: () => s.strings.error_bitpay_empty_output_invoice

File: src/constants/WalletAndCurrencyConstants.js (hunk 6 above)
Comment (peachbits) [inline src/constants/WalletAndCurrencyConstants.js:367]: Should be false

File: src/types/BitPayError.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  [BitPayErrorCode.EmptyVerificationHexReq]: (params: { uri: string }) => s.strings.error_bitpay_empty_verification_hex_req,
+  [BitPayErrorCode.FetchFailed]: (params: { uri: string, body: string, header: string, method: string, statusCode: string, text: string }) =>
```
Comment (peachbits) [inline src/types/BitPayError.js:36]: [BitPayErrorCode.FetchFailed]: (params: { header: string, statusCode: string, text: string }) =>  sprintf(s.strings.error_bitpay_fetch, params.header, params.statusCode, params.text),

**Review by peachbits — CHANGES_REQUESTED** [review]

**Review by peachbits — APPROVED** [review]

**Pull request merged**